### PR TITLE
PV2-404: clear required payments cache between e2e tests

### PR DESCRIPTION
**Pull request by kuzinator**
... and a few other fixes to revealed false positives

### CODE REVIEW

File: src/SFA.DAS.Payments.AcceptanceTests.EndToEnd/Steps/EndToEndSteps.cs [hunk 1]
```diff
@@ -51,6 +53,17 @@ public async Task ThenTheFollowingLearnerEarningsShouldBeGenerated(Table table)
...
 
+                //var testLearner = TestSession.Learners.FirstOrDefault(l => l.LearnerIdentifier == learnerId);
+                //if (testLearner == null) TestSession.Learners.Add(new Learner
+                //{
+                //    LearnerIdentifier = learnerId,
+                //    LearnRefNumber = learner.LearnRefNumber,
+                //    Course = new Course
+                //    {
+                //        LearnAimRef = learner.LearningDeliveries[0].LearningDeliveryValues.LearnAimRef
+                //    }
```
Comment (samuel-dev-dba) [inline src/SFA.DAS.Payments.AcceptanceTests.EndToEnd/Steps/EndToEndSteps.cs]: remove code comment if not needed

File: src/SFA.DAS.Payments.RequiredPayments.RequiredPaymentsService/RequiredPaymentsService.cs [hunk 2]
```diff
@@ -88,18 +90,21 @@ public async Task Initialise()
...
-            await StateManager.AddStateAsync("initialised", false).ConfigureAwait(false);
+            if (await IsInitialised().ConfigureAwait(false))
+                await StateManager.RemoveStateAsync(InitialisedKey, CancellationToken.None).ConfigureAwait(false);
         }
 
         private async Task<bool> IsInitialised()
         {
-            return await StateManager.ContainsStateAsync("initialised").ConfigureAwait(false);
+            var isInitialised = await StateManager.ContainsStateAsync(InitialisedKey).ConfigureAwait(false);
+            Debug.WriteLine("isInitialised: " + isInitialised);
```
Comment (samuel-dev-dba) [inline src/SFA.DAS.Payments.RequiredPayments.RequiredPaymentsService/RequiredPaymentsService.cs]: please write this to log

File: src/SFA.DAS.Payments.AcceptanceTests.ServiceFabric.Core/Class1.cs [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+        //{
+        //        var actorId = new ActorId(key);
+        //        return proxyFactory.CreateActorProxy<T>(new Uri(url), actorId);
+
+        //    foreach (var learner in testSession.Learners)
+        //    {
+        //        //var key = apprenticeshipKeyService.GenerateApprenticeshipKey(
+        //        //    testSession.Ukprn,
+        //        //    testSession.GetLearner(learner.LearnerIdentifier).LearnerIdentifier,
+        //        //    learner.Course.FrameworkCode,
```
Comment (samuel-dev-dba) [inline src/SFA.DAS.Payments.AcceptanceTests.ServiceFabric.Core/Class1.cs]: Please remove the commented code

File: src/SFA.DAS.Payments.AcceptanceTests.EndToEnd/Steps/EndToEndStepsBase.cs [hunk 4]
```diff
@@ -243,17 +250,52 @@ private static List<PriceEpisode> GetPriceEpisodes(Training learnerEarnings, Pri
...
-                priceEpisode.PriceEpisodePeriodisedValues.Add(learningValues);
-                priceEpisode.PriceEpisodePeriodisedValues.Add(completionEarnings);
-                priceEpisode.PriceEpisodePeriodisedValues.Add(balancingEarnings);
+                PriceEpisodePeriodisedValues episodeLearningValues;
+                PriceEpisodePeriodisedValues episodeCompletionEarnings;
+                PriceEpisodePeriodisedValues episodeBalancingEarnings;
+
+                if ((episodeStart.Period > 1 && episodeStart.GetCollectionYear() == collectionYear) || episodeLastPeriod < 12)
+                {
+                    episodeLearningValues = new PriceEpisodePeriodisedValues {AttributeName = learningValues.AttributeName};
```
Comment (samuel-dev-dba) [inline src/SFA.DAS.Payments.AcceptanceTests.EndToEnd/Steps/EndToEndStepsBase.cs]: Why do we have to change this ?

File: src/SFA.DAS.Payments.AcceptanceTests.EndToEnd/Steps/EndToEndSteps.cs (hunk 1 above)
Comment (kuzinator) [reply src/SFA.DAS.Payments.AcceptanceTests.EndToEnd/Steps/EndToEndSteps.cs]: done

File: src/SFA.DAS.Payments.RequiredPayments.RequiredPaymentsService/RequiredPaymentsService.cs (hunk 2 above)
Comment (kuzinator) [reply src/SFA.DAS.Payments.RequiredPayments.RequiredPaymentsService/RequiredPaymentsService.cs]: This was a leftover. In fact, i've inlined this method altogether now as it's never used without call to Initialise.

File: src/SFA.DAS.Payments.AcceptanceTests.ServiceFabric.Core/Class1.cs (hunk 3 above)
Comment (kuzinator) [reply src/SFA.DAS.Payments.AcceptanceTests.ServiceFabric.Core/Class1.cs]: this project shouldn't be there at all. removed

File: src/SFA.DAS.Payments.AcceptanceTests.EndToEnd/Steps/EndToEndStepsBase.cs (hunk 4 above)
Comment (kuzinator) [reply src/SFA.DAS.Payments.AcceptanceTests.EndToEnd/Steps/EndToEndStepsBase.cs]: because with multiple price episodes we only need to populate values relevant to each episode, otherwise it was generating extra earnings for price episodes that haven't started or already finished

File: src/SFA.DAS.Payments.RequiredPayments.RequiredPaymentsService/RequiredPaymentsService.cs (hunk 2 above)
Comment (samuel-dev-dba) [reply src/SFA.DAS.Payments.RequiredPayments.RequiredPaymentsService/RequiredPaymentsService.cs]: Debug.WriteLine("isInitialised: " + isInitialised); should be logged

File: src/SFA.DAS.Payments.RequiredPayments.RequiredPaymentsService.Interfaces/SFA.DAS.Payments.RequiredPayments.RequiredPaymentsService.Interfaces.csproj [hunk 5]
```diff
@@ -1,10 +1,12 @@
...
 
   <PropertyGroup>
-    <TargetFramework>netcoreapp2.1</TargetFramework>
-    <RuntimeFrameworkVersion>2.1.2</RuntimeFrameworkVersion>
+    <TargetFramework>netstandard2.0</TargetFramework>
     <IsServiceFabricServiceProject>True</IsServiceFabricServiceProject>
-    <RuntimeIdentifier>win7-x64</RuntimeIdentifier>
+  </PropertyGroup>
+
+  <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Debug|AnyCPU'">
```
Comment (samuel-dev-dba) [inline src/SFA.DAS.Payments.RequiredPayments.RequiredPaymentsService.Interfaces/SFA.DAS.Payments.RequiredPayments.RequiredPaymentsService.Interfaces.csproj]: why the change this may lead to a deployment issue

File: src/SFA.DAS.Payments.RequiredPayments.RequiredPaymentsService.Interfaces/SFA.DAS.Payments.RequiredPayments.RequiredPaymentsService.Interfaces.csproj [hunk 6]
```diff
@@ -1,10 +1,12 @@
 ﻿<Project Sdk="Microsoft.NET.Sdk">
 
   <PropertyGroup>
-    <TargetFramework>netcoreapp2.1</TargetFramework>
-    <RuntimeFrameworkVersion>2.1.2</RuntimeFrameworkVersion>
+    <TargetFramework>netstandard2.0</TargetFramework>
```
Comment (samuel-dev-dba) [inline src/SFA.DAS.Payments.RequiredPayments.RequiredPaymentsService.Interfaces/SFA.DAS.Payments.RequiredPayments.RequiredPaymentsService.Interfaces.csproj]: please why the change ?

File: src/SFA.DAS.Payments.AcceptanceTests.Core/StepsBase.cs [hunk 7]
```diff
@@ -50,10 +56,16 @@ protected async Task WaitForIt(Func<Task<bool>> lookForIt, string failText)
...
+            while (DateTime.Now < endTime || lastRun)
             {
                 if (lookForIt())
-                    return;
+                {
+                    if (lastRun) return;
+                    lastRun = true;
+                }
+
                 await Task.Delay(Config.TimeToPause);
```
Comment (carlbm) [inline src/SFA.DAS.Payments.AcceptanceTests.Core/StepsBase.cs]: Does it make sense to introduce a local method here to reduce the duplication with the method on lines 43-50?

File: src/SFA.DAS.Payments.AcceptanceTests.EndToEnd/EventMatchers/RequiredPaymentEventMatcher.cs [hunk 8]
```diff
@@ -58,13 +58,13 @@ protected override IList<ApprenticeshipContractType2RequiredPaymentEvent> GetExp
...
                     expectedPayments.Add(learningPayment);
 
-                if (balancingPayment.AmountDue > 0) 
+                if (balancingPayment.AmountDue != 0) 
                     expectedPayments.Add(balancingPayment);
 
-                if (completionPayment.AmountDue > 0) 
+                if (completionPayment.AmountDue != 0) 
                     expectedPayments.Add(completionPayment);
             }
```
Comment (carlbm) [inline src/SFA.DAS.Payments.AcceptanceTests.EndToEnd/EventMatchers/RequiredPaymentEventMatcher.cs]: Are these checks correct?  I would imagine that you'd want to add all payments as payments where the value = 0 are probably incorrect anyway

File: src/SFA.DAS.Payments.AcceptanceTests.EndToEnd/EventMatchers/RequiredPaymentEventMatcher.cs [hunk 9]
```diff
@@ -58,13 +58,13 @@ protected override IList<ApprenticeshipContractType2RequiredPaymentEvent> GetExp
                     DeliveryPeriod = payment.DeliveryPeriod.ToCalendarPeriod()
                 };
```
Comment (carlbm) [inline src/SFA.DAS.Payments.AcceptanceTests.EndToEnd/EventMatchers/RequiredPaymentEventMatcher.cs]: Does it make sense to refactor this method to loop through all values in the Enum? Going by the rule of 3 it's time to simplify it

File: src/SFA.DAS.Payments.AcceptanceTests.EndToEnd/Steps/EndToEndStepsBase.cs [hunk 10]
```diff
@@ -243,17 +250,52 @@ private static List<PriceEpisode> GetPriceEpisodes(Training learnerEarnings, Pri
...
+                    episodeCompletionEarnings = new PriceEpisodePeriodisedValues { AttributeName = completionEarnings.AttributeName };
+                    episodeBalancingEarnings = new PriceEpisodePeriodisedValues { AttributeName = balancingEarnings.AttributeName };
+
+                    for (var p = 1; p < 13; p++)
+                    {
+                        var propertyInfo = typeof(PriceEpisodePeriodisedValues).GetProperty("Period" + p);
+
```
Comment (carlbm) [inline src/SFA.DAS.Payments.AcceptanceTests.EndToEnd/Steps/EndToEndStepsBase.cs]: Should we use fastmember?

Comment (kuzinator) [reply src/SFA.DAS.Payments.AcceptanceTests.EndToEnd/Steps/EndToEndStepsBase.cs]: We might when volume is right or performance is wrong or it's a production code.

File: src/SFA.DAS.Payments.AcceptanceTests.Core/StepsBase.cs (hunk 7 above)
Comment (kuzinator) [reply src/SFA.DAS.Payments.AcceptanceTests.Core/StepsBase.cs]: I thought about it. They're all slightly different and trying to factor our common parts would only add to awkwardness. Happy to be wrong on this one though.

File: src/SFA.DAS.Payments.AcceptanceTests.EndToEnd/EventMatchers/RequiredPaymentEventMatcher.cs (hunk 8 above)
Comment (kuzinator) [reply src/SFA.DAS.Payments.AcceptanceTests.EndToEnd/EventMatchers/RequiredPaymentEventMatcher.cs]: RequiredPayments component doesn't produce £0 payments hence this code.

File: src/SFA.DAS.Payments.AcceptanceTests.EndToEnd/EventMatchers/RequiredPaymentEventMatcher.cs (hunk 9 above)
Comment (kuzinator) [reply src/SFA.DAS.Payments.AcceptanceTests.EndToEnd/EventMatchers/RequiredPaymentEventMatcher.cs]: We're adding incentives this sprint, the number will grow and ask for refactoing.
